x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java

```diff
@@ -392,33 +404,73 @@ protected void doRun() throws Exception {
                                         TimeValue.timeValueNanos(statsCurrentTimeNanosSupplier.getAsLong() - startTimeInNanos).millis()
                                     )
                                 );
+                            }));
+                        }
+                        enqueued = true;
+                    } else {
+                        logger.warn("{} too many files ({}) to warm in cache, skipping file [{}]", shardId, queue.size(), fileName);
+                    }
+                } catch (Exception e) {
+                    logger.warn(
+                        () -> new ParameterizedMessage("unable to open file [{}], cancelling prewarming for shard [{}]", fileName, shardId),
+                        e
+                    );
+                    if (queue.size() > 0) {
+                        try {
+                            ActionRunnable<Void> warmer;
+                            while ((warmer = queue.poll(0L, TimeUnit.MILLISECONDS)) != null) {
+                                warmer.onFailure(e);
                             }
+                        } catch (InterruptedException ie) {
+                            Thread.currentThread().interrupt();
+                            logger.warn(() -> new ParameterizedMessage("{} shard cache warming has been interrupted", shardId), ie);
+                        }
+                    }
+                    if (e instanceof AlreadyClosedException || (e.getCause() != null && e.getCause() instanceof AlreadyClosedException)) {
+                        break;
+                    } else {
+                        throw new ElasticsearchException("Exception when warming cache for shard " + shardId, e);
+                    }
+                } finally {
+                    if (enqueued == false) {
+                        IOUtils.closeWhileHandlingException(input);
+                    }
+                }
+            }
 
-                            @Override
-                            public void onFailure(Exception e) {
-                                logger.trace(
-                                    () -> new ParameterizedMessage(
-                                        "failed to warm cache for [{}] part [{}/{}]",
-                                        fileName,
-                                        part,
-                                        numberOfParts
-                                    ),
-                                    e
-                                );
-                            }
+            // Start as many workers as fit into the searchable snapshot pool at once at the most
+            final int workers = Math.min(threadPool.info(SEARCHABLE_SNAPSHOTS_THREAD_POOL_NAME).getMax(), queue.size());
+            for (int i = 0; i < workers; ++i) {
```

Comment:
Why do our own queuing here? Can't we just push all the tasks in `queue` onto the `executor` and let it poll tasks from its internal queue when we're using the full pool anyway?